kernel/sem.c

```diff
@@ -98,18 +113,165 @@ int z_vrfy_k_sem_init(struct k_sem *sem, unsigned int initial_count,
 static inline void handle_poll_events(struct k_sem *sem)
 {
 #ifdef CONFIG_POLL
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+	k_spinlock_key_t key = k_spin_lock(&poll_lock);
+
+#endif
 	z_handle_obj_poll_events(&sem->poll_events, K_POLL_STATE_SEM_AVAILABLE);
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+	k_spin_unlock(&poll_lock, key);
+#endif
 #else
 	ARG_UNUSED(sem);
 #endif
 }
 
-void z_impl_k_sem_give(struct k_sem *sem)
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+/*
+ * Quick overview of this semaphore implementation:
+ * `count` is always eventually equal to `limit`, minus the number
+ * of outstanding `sem_take`s, minus the number of waiting threads
+ * in `wait_q`.
+ *
+ * Once a thread has incremented or decremented `count`, it is
+ * responsible for ensuring the above is promptly restored.
+ *
+ */
+
+static void sem_give(struct k_sem *sem)
+{
+	atomic_t count;
+	atomic_t limit_minus_one = sem->limit - 1;
```

Comment:
Style: I feel comfortable that a expression including a variable named "limit_minus_one" can be very reasonably expressed using "(limit - 1)" without loss of clarity or abstractive value. :)

Reply:
The compiler was being annoying, and was doing the decrement inside the CAS loop...

Reply:
Ah, OK. Then maybe a "/* keep subtract outside of lock */" comment would be helpful so no one comes by to "fix" this.

Reply:
Sounds good...

- [ ] Add comment to prevent driveby "fixes".